File: scripts/ingest-openprescribing.mjs
import "./_env.mjs";
/*
  Atlas-style OpenPrescribing ingest for the QOF Optimiser.

  Pulls mean-monthly prescribing ITEMS per BNF drug group from OpenPrescribing's
  spending_by_org endpoint (one call per group per month — NOT per practice),
  converts to items per 1,000 registered patients, rolls up practice -> PCN ->
  ICB -> England (list-weighted), computes a percentile within each level, and
  writes rx_value. Mirrors the CVD & Diabetes atlas ETL.

  Run on your machine (needs internet + service-role key):
      node scripts/ingest-openprescribing.mjs
  Options:
      MONTHS=12   how many recent months to average over (default 12)
      DRY=1       fetch + compute, print a summary, but don't write

  Env: NEXT_PUBLIC_SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY
*/
import { createClient } from "@supabase/supabase-js";

const supa = createClient(process.env.NEXT_PUBLIC_SUPABASE_URL, process.env.SUPABASE_SERVICE_ROLE_KEY);
const BASE = "https://openprescribing.net/api/1.0";
const MONTHS = Number(process.env.MONTHS || 12);
const DRY = process.env.DRY === "1";

// ---- metric -> BNF chemical codes (fallbacks). Single-substance drugs are also
//      re-resolved by NAME at runtime because BNF codes drift over time. ----
const GROUPS = {
  rx_sglt2i: {
    codes: ["0601023AG", "0601023AN", "0601023AM", "0601023AX",
            "0601023AL", "0601023AP", "0601023AR", "0601023AY", "0601023AV"],
    names: ["dapagliflozin", "empagliflozin", "canagliflozin", "ertugliflozin"],
  },
  rx_glp1_sema:  { codes: ["0601023AW"], names: ["semaglutide"] },
  rx_tirzepatide:{ codes: ["0601023AZ"], names: ["tirzepatide"] },
  rx_statin:     { codes: ["0212000"],   names: [] },              // whole lipid-regulating section
  rx_ezetimibe:  { codes: ["0212000L0"], names: ["ezetimibe"] },
  rx_inclisiran: { codes: ["0212000AM"], names: ["inclisiran"] },
  rx_doac:       { codes: ["0208020Z0", "0208020Y0", "0208020AA", "0208020X0"],
                   names: ["apixaban", "rivaroxaban", "edoxaban", "dabigatran"] },
};
// high-intensity statin presentations: atorvastatin 20/40/80, rosuvastatin 10/20/40
const HI_STATIN = { atorvastatin: [20, 40, 80], rosuvastatin: [10, 20, 40] };

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

const UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36";

// OpenPrescribing gates its API behind a Cloudflare JS challenge that plain HTTP
// clients (fetch/requests) cannot pass, but a real browser engine solves it
// automatically. So we drive a headless Chromium via Playwright and read the JSON
// straight off the page. One shared page carries the cf_clearance cookie across
// all calls, so only the first request pays the challenge cost.
let _page = null;
let _browser = null;

async function initFetcher() {
  let chromium;
  try {
    ({ chromium } = await import("playwright"));
  } catch {
    console.error("\nThis ingest needs Playwright (a headless browser) to get past OpenPrescribing's Cloudflare check.\nInstall it once, then re-run:\n\n    npm install -D playwright\n    npx playwright install chromium\n    node scripts/ingest-openprescribing.mjs\n");
    process.exit(1);
  }
  _browser = await chromium.launch({ headless: true });
  const ctx = await _browser.newContext({ userAgent: UA });
  _page = await ctx.newPage();
}

async function closeFetcher() {
  try { await _browser?.close(); } catch { /* ignore */ }
}

async function api(path, params) {
  const qs = new URLSearchParams({ ...params, format: "json" }).toString();
  const url = `${BASE}${path}?${qs}`;
  let lastReason = "unknown";
  for (let attempt = 0; attempt < 4; attempt++) {
    try {
      await _page.goto(url, { waitUntil: "domcontentloaded", timeout: 90000 });
      // Cloudflare may briefly show an interstitial; poll until the body is JSON.
      for (let i = 0; i < 20; i++) {
        const text = await _page.evaluate(() => document.body?.innerText || "");
        const t = text.trim();
        if (t.startsWith("[") || t.startsWith("{")) {
          try { return JSON.parse(t); } catch { /* still rendering */ }
        }
        lastReason = `non-JSON body: ${t.slice(0, 80).replace(/\s+/g, " ")}`;
        await sleep(1500);
      }
    } catch (e) {
      lastReason = `${e.name}: ${e.message}`;
    }
    await sleep(1500 * (attempt + 1));
  }
  throw new Error(`GET ${url}\n  reason: ${lastReason}`);
}

function windowMonths(latest, n) {
  // n YYYY-MM-01 strings ending at `latest` (inclusive).
  const [y, m] = [Number(latest.slice(0, 4)), Number(latest.slice(5, 7))];
  const out = [];
  for (let i = n - 1; i >= 0; i--) {
    let yy = y, mm = m - i;
    while (mm <= 0) { mm += 12; yy -= 1; }
    out.push(`${yy}-${String(mm).padStart(2, "0")}-01`);
  }
  return out;
}

async function latestMonth(code) {
  const rows = await api("/spending/", { code });
  const dates = (rows || []).map((r) => r.date).filter(Boolean);
  return dates.length ? dates.sort().at(-1) : null;
}

// {practice_code: mean monthly items} for a single BNF code across the window.
async function meanItemsByPractice(code, months) {
  const totals = {};
  let published = 0;
  for (const date of months) {
    let rows;
    try { rows = await api("/spending_by_org/", { org_type: "practice", code, date }); }
    catch { rows = null; }
    if (rows && rows.length) {
      published++;
      for (const r of rows) if (r.row_id) totals[r.row_id] = (totals[r.row_id] || 0) + (r.items || 0);
    }
    await sleep(120);
  }
  if (!published) return {};
  const out = {};
  for (const [c, v] of Object.entries(totals)) out[c] = v / published;
  return out;
}

// Resolve single-substance drug names to current BNF chemical codes (drift-proof).
async function resolveNames(names) {
  const out = [];
  for (const name of names) {
    try {
      const rows = await api("/bnf_code/", { q: name });
      const chem = (rows || []).filter((r) => String(r.type || "").toLowerCase() === "chemical");
      const nl = name.toLowerCase();
      const pick = chem.find((r) => String(r.name || "").toLowerCase() === nl)
        || chem.find((r) => String(r.name || "").toLowerCase().includes(nl) && !String(r.name || "").includes("/"))
        || chem[0];
      if (pick?.id) out.push(pick.id);
    } catch { /* skip */ }
  }
  return out;
}

// Presentation codes for the high-intensity statin strengths, resolved by name.
async function hiStatinCodes() {
  const codes = [];
  for (const [drug, strengths] of Object.entries(HI_STATIN)) {
    let rows;
    try { rows = await api("/bnf_code/", { q: drug }); } catch { rows = []; }
    for (const r of rows || []) {
      if (String(r.type || "").toLowerCase() !== "presentation") continue;
      const nm = String(r.name || "").toLowerCase();
      if (nm.includes("/")) continue; // exclude combinations
      if (strengths.some((s) => nm.includes(`${s}mg`)) && r.id) codes.push(r.id);
    }
  }
  return codes;
}

function percentiles(rateByCode) {
  // 0-100 percentile of each value within the set (ties share the lower rank).
  const vals = Object.values(rateByCode).filter((v) => v != null).sort((a, b) => a - b);
  const n = vals.length;
  const out = {};
  for (const [code, v] of Object.entries(rateByCode)) {
    if (v == null) { out[code] = null; continue; }
    let below = 0; while (below < n && vals[below] < v) below++;
    out[code] = n > 1 ? Math.round((below / (n - 1)) * 100) : 50;
  }
  return out;
}

async function main() {
  console.log(`OpenPrescribing ingest — ${MONTHS}-month mean${DRY ? " (DRY RUN)" : ""}`);
  await initFetcher();
  console.log("  headless browser ready (getting past Cloudflare)…");

  // organisation hierarchy + list sizes (paged past the 1000-row cap)
  const orgs = [];
  for (let from = 0; ; from += 1000) {
    const { data, error } = await supa.from("organisation")
      .select("ods_code,org_level,parent_pcn,parent_icb,list_size").order("ods_code").range(from, from + 999);
    if (error) throw error;
    if (!data?.length) break;
    orgs.push(...data);
    if (data.length < 1000) break;
  }
  const practices = orgs.filter((o) => o.org_level === "practice");
  const listByCode = Object.fromEntries(orgs.map((o) => [o.ods_code, o.list_size || 0]));
  const parentPcn = Object.fromEntries(practices.map((p) => [p.ods_code, p.parent_pcn]));
  const parentIcb = Object.fromEntries(practices.map((p) => [p.ods_code, p.parent_icb]));
  console.log(`  ${practices.length} practices, ${orgs.length} orgs`);

  // period anchor: latest published month for a common code
  const latest = (await latestMonth("0212000")) || (await latestMonth("0601023AG"));
  if (!latest) throw new Error("Could not determine latest OpenPrescribing month.");
  const months = windowMonths(latest, MONTHS);
  const period = `${months[0].slice(0, 7)}..${months.at(-1).slice(0, 7)} (mean monthly)`;
  console.log(`  window: ${period}`);

  // resolve high-intensity statin presentations once
  const hiCodes = await hiStatinCodes();
  console.log(`  high-intensity statin presentations resolved: ${hiCodes.length}`);

  // raw mean-monthly items per practice, per metric
  const rawItems = {}; // metric_key -> {practice: items}
  for (const [key, grp] of Object.entries(GROUPS)) {
    let codes = grp.codes;
    if (grp.names.length) {
      const resolved = await resolveNames(grp.names);
      if (resolved.length) codes = resolved;
    }
    const agg = {};
    for (const c of codes) {
      const m = await meanItemsByPractice(c, months);
      for (const [pc, items] of Object.entries(m)) agg[pc] = (agg[pc] || 0) + items;
      await sleep(150);
    }
    rawItems[key] = agg;
    console.log(`  ${key}: ${Object.keys(agg).length} practices with items`);
  }

  // high-intensity statin items (numerator for the share)
  const hiItems = {};
  for (const c of hiCodes) {
    const m = await meanItemsByPractice(c, months);
    for (const [pc, items] of Object.entries(m)) hiItems[pc] = (hiItems[pc] || 0) + items;
    await sleep(120);
  }

  // ---- build rows: practice rates, roll-ups, percentiles ----
  const rows = [];
  const metricKeys = [...Object.keys(GROUPS), "rx_statin_hi"];

  for (const key of metricKeys) {
    const isShare = key === "rx_statin_hi";
    // practice-level rate
    const practiceRate = {}; const practiceRaw = {};
    for (const p of practices) {
      const code = p.ods_code, list = listByCode[code] || 0;
      if (isShare) {
        const denom = rawItems.rx_statin[code] || 0;
        practiceRaw[code] = hiItems[code] || 0;
        practiceRate[code] = denom > 0 ? Math.round((hiItems[code] || 0) / denom * 1000) / 10 : null;
      } else {
        const items = rawItems[key][code];
        practiceRaw[code] = items ?? null;
        practiceRate[code] = items != null && list > 0 ? Math.round((items / list * 1000) * 100) / 100 : (items != null ? 0 : null);
      }
    }
    const pPct = percentiles(practiceRate);
    for (const p of practices) {
      if (practiceRate[p.ods_code] == null) continue;
      rows.push({ ods_code: p.ods_code, org_level: "practice", metric_key: key, period,
        raw_items: practiceRaw[p.ods_code], items_per_1000: practiceRate[p.ods_code], percentile: pPct[p.ods_code] });
    }

    // roll up to PCN / ICB / England (list-weighted). Shares use hi/statin item sums.
    for (const [level, parentOf] of [["pcn", parentPcn], ["icb", parentIcb]]) {
      const num = {}, den = {}, hi = {}, stat = {};
      for (const p of practices) {
        const parent = parentOf[p.ods_code]; if (!parent) continue;
        if (isShare) {
          hi[parent] = (hi[parent] || 0) + (hiItems[p.ods_code] || 0);
          stat[parent] = (stat[parent] || 0) + (rawItems.rx_statin[p.ods_code] || 0);
        } else if (rawItems[key][p.ods_code] != null) {
          num[parent] = (num[parent] || 0) + rawItems[key][p.ods_code];
          den[parent] = (den[parent] || 0) + (listByCode[p.ods_code] || 0);
        }
      }
      const rate = {};
      const codes = isShare ? Object.keys(stat) : Object.keys(num);
      for (const c of codes) {
        rate[c] = isShare
          ? (stat[c] > 0 ? Math.round(hi[c] / stat[c] * 1000) / 10 : null)
          : (den[c] > 0 ? Math.round((num[c] / den[c] * 1000) * 100) / 100 : null);
      }
      const pct = percentiles(rate);
      for (const c of codes) {
        if (rate[c] == null) continue;
        rows.push({ ods_code: c, org_level: level, metric_key: key, period,
          raw_items: isShare ? hi[c] : num[c], items_per_1000: rate[c], percentile: pct[c] });
      }
    }

    // England
    let engRate, engRaw;
    if (isShare) {
      const hiSum = practices.reduce((s, p) => s + (hiItems[p.ods_code] || 0), 0);
      const statSum = practices.reduce((s, p) => s + (rawItems.rx_statin[p.ods_code] || 0), 0);
      engRaw = hiSum; engRate = statSum > 0 ? Math.round(hiSum / statSum * 1000) / 10 : null;
    } else {
      const numSum = practices.reduce((s, p) => s + (rawItems[key][p.ods_code] || 0), 0);
      const listSum = practices.reduce((s, p) => s + (rawItems[key][p.ods_code] != null ? (listByCode[p.ods_code] || 0) : 0), 0);
      engRaw = numSum; engRate = listSum > 0 ? Math.round((numSum / listSum * 1000) * 100) / 100 : null;
    }
    if (engRate != null) rows.push({ ods_code: "ENG", org_level: "national", metric_key: key, period, raw_items: engRaw, items_per_1000: engRate, percentile: 50 });
  }

  console.log(`  built ${rows.length} rows`);
  if (DRY) {
    const eng = rows.filter((r) => r.org_level === "national");
    console.log("  England rates:", Object.fromEntries(eng.map((r) => [r.metric_key, r.items_per_1000])));
    console.log("DRY RUN — nothing written.");
    return;
  }

  // write in chunks
  for (let i = 0; i < rows.length; i += 500) {
    const chunk = rows.slice(i, i + 500);
    const { error } = await supa.from("rx_value").upsert(chunk);
    if (error) { console.error("upsert error", error.message); process.exit(1); }
  }
  console.log(`Prescribing ingest complete — ${rows.length} rows for ${period}.`);
}

main()
  .catch((e) => { console.error(e); process.exitCode = 1; })
  .finally(closeFetcher);

File: scripts/ingest_openprescribing.py
"""
OpenPrescribing -> QOF Optimiser prescribing ingest (Python / requests).

Uses the same plain-requests method as the working Pydroid script. Fetches the
8 app measures (adds DOACs), converts to mean monthly items per 1,000 patients,
rolls up practice -> PCN -> ICB -> England with a peer percentile, and upserts
into the app's rx_value table via the Supabase REST API.

Run from the qof-optimiser folder:
    python scripts/ingest_openprescribing.py            # full run
    python scripts/ingest_openprescribing.py --test     # connectivity test only
    python scripts/ingest_openprescribing.py --dry      # fetch + compute, don't write
    MONTHS=3 python scripts/ingest_openprescribing.py    # window length (default 3)

Needs: pip install requests   (pandas NOT required)
Reads NEXT_PUBLIC_SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY from .env.local
"""
import os, sys, time, math, json
import requests

BASE = "https://openprescribing.net/api/1.0"
DRY = "--dry" in sys.argv
TEST_ONLY = "--test" in sys.argv
N_MONTHS = int(os.environ.get("MONTHS", "3"))
SLEEP = 0.25

session = requests.Session()
session.headers.update({"User-Agent": "GP-Prescribing-Analysis/1.0 (research)"})

# metric_key -> BNF chemical codes to sum (verified against openprescribing.net/bnf/)
MEASURES = {
    "rx_sglt2i": ["0601023AG", "0601023AN", "0601023AM", "0601023AX",
                  "0601023AL", "0601023AP", "0601023AR", "0601023AY", "0601023AV"],
    "rx_glp1_sema": ["0601023AW"],
    "rx_tirzepatide": ["0601023AZ"],
    "rx_ezetimibe": ["0212000L0"],
    "rx_inclisiran": ["0212000AM"],
    "rx_doac": ["0208020Z0", "0208020Y0", "0208020AA", "0208020X0"],  # apixaban, rivaroxaban, edoxaban, dabigatran
}
STATIN_CHEMS = ["0212000B0", "0212000AA", "0212000Y0", "0212000X0", "0212000M0"]  # ator, rosu, simva, prava, fluva
HI_STATINS = {"0212000B0": ["20mg", "40mg", "80mg"], "0212000AA": ["10mg", "20mg", "40mg"]}  # atorvastatin, rosuvastatin

# Ratio (%) measures: numerator items / denominator items * 100 (higher = better).
# rx_doac_anticoag: DOACs as a share of oral anticoagulants (DOAC vs warfarin).
DOAC_CODES = ["0208020Z0", "0208020Y0", "0208020AA", "0208020X0"]  # apixaban, rivaroxaban, edoxaban, dabigatran
WARFARIN_CODES = ["0208020V0"]  # warfarin sodium
# rx_metformin_mr: modified-release metformin as a share of all metformin.
METFORMIN_CHEM = "0601022B0"
MR_NAME_HINTS = ["m/r", "modified", " sr ", " sr", "prolonged", "xr", " mr "]  # MR presentation name markers


def api_get(path, params=None, retries=3):
    params = dict(params or {}); params["format"] = "json"
    last = None
    for attempt in range(retries):
        try:
            r = session.get(f"{BASE}{path}", params=params, timeout=120)
            if r.status_code == 200:
                return r.json()
            last = f"HTTP {r.status_code}: {r.text[:120].strip()}"
        except requests.RequestException as e:
            last = f"{type(e).__name__}: {e}"
        time.sleep(2 ** attempt)
    raise RuntimeError(f"GET {path} failed — {last}")


def load_env():
    env = {}
    path = os.path.join(os.getcwd(), ".env.local")
    if os.path.exists(path):
        for line in open(path, encoding="utf-8"):
            line = line.strip()
            if line and not line.startswith("#") and "=" in line:
                k, v = line.split("=", 1)
                env[k.strip()] = v.strip()
    return env


def latest_month(code):
    rows = api_get("/spending/", {"code": code}) or []
    dates = [r.get("date") for r in rows if r.get("date")]
    return max(dates) if dates else None


def window_months(latest, n):
    y, m = int(latest[:4]), int(latest[5:7])
    out = []
    for i in range(n - 1, -1, -1):
        yy, mm = y, m - i
        while mm <= 0:
            mm += 12; yy -= 1
        out.append(f"{yy:04d}-{mm:02d}-01")
    return out


def mean_items_by_practice(code, months):
    """{practice_code: mean monthly items} for a BNF code, GP practices only."""
    totals, published = {}, 0
    for month in months:
        rows = api_get("/spending_by_org/", {"org_type": "practice", "code": code, "date": month}) or []
        if rows:
            published += 1
            for r in rows:
                if r.get("row_id") and (r.get("setting") in (4, None)):
                    totals[r["row_id"]] = totals.get(r["row_id"], 0) + (r.get("items") or 0)
        time.sleep(SLEEP)
    return {c: v / published for c, v in totals.items()} if published else {}


def hi_statin_presentation_codes():
    codes = []
    for chem, strengths in HI_STATINS.items():
        try:
            items = api_get("/bnf_code/", {"q": chem})
        except Exception:
            continue
        for it in items or []:
            bid, name = it.get("id", ""), (it.get("name") or "").lower()
            if len(bid) == 15 and bid.startswith(chem) and "/" not in name and any(s in name for s in strengths):
                codes.append(bid)
        time.sleep(SLEEP)
    return codes


def sb_get_all(url, key, table, select):
    out, offset = [], 0
    while True:
        r = requests.get(f"{url}/rest/v1/{table}", params={"select": select, "limit": 1000, "offset": offset},
                         headers={"apikey": key, "Authorization": f"Bearer {key}"}, timeout=60)
        r.raise_for_status()
        chunk = r.json()
        out.extend(chunk)
        if len(chunk) < 1000:
            break
        offset += 1000
    return out


def sb_upsert(url, key, table, rows):
    for i in range(0, len(rows), 500):
        chunk = rows[i:i + 500]
        r = requests.post(f"{url}/rest/v1/{table}",
                          headers={"apikey": key, "Authorization": f"Bearer {key}",
                                   "Content-Type": "application/json", "Prefer": "resolution=merge-duplicates"},
                          data=json.dumps(chunk), timeout=120)
        if r.status_code >= 300:
            raise RuntimeError(f"upsert failed HTTP {r.status_code}: {r.text[:200]}")


def percentiles(rate):
    vals = sorted(v for v in rate.values() if v is not None)
    n = len(vals)
    out = {}
    for c, v in rate.items():
        if v is None:
            out[c] = None; continue
        below = sum(1 for x in vals if x < v)
        out[c] = round(below / (n - 1) * 100) if n > 1 else 50
    return out


def main():
    # ---- 0. connectivity test ----
    print("Testing OpenPrescribing reachability...")
    try:
        latest = latest_month("0212000B0")
        print(f"  OK — reachable. Latest published month: {latest}")
    except Exception as e:
        print(f"  BLOCKED — {e}\n\nThis network can't reach OpenPrescribing (same Cloudflare wall as before).")
        print("Try another network, or run the fetch where it works and import the CSV instead.")
        sys.exit(1)
    if TEST_ONLY:
        print("\n--test: connectivity confirmed, stopping here.")
        return

    months = window_months(latest, N_MONTHS)
    period = f"{months[0][:7]}..{months[-1][:7]} (OpenPrescribing, mean monthly)"
    print(f"Window: {period}\n")

    env = load_env()
    url, key = env.get("NEXT_PUBLIC_SUPABASE_URL"), env.get("SUPABASE_SERVICE_ROLE_KEY")
    if not url or not key:
        print("Missing NEXT_PUBLIC_SUPABASE_URL / SUPABASE_SERVICE_ROLE_KEY in .env.local"); sys.exit(1)

    # ---- 1. org hierarchy + list sizes (from the app DB) ----
    orgs = sb_get_all(url, key, "organisation", "ods_code,org_level,parent_pcn,parent_icb,list_size")
    practices = [o for o in orgs if o["org_level"] == "practice"]
    list_by = {o["ods_code"]: (o.get("list_size") or 0) for o in orgs}
    pcn_of = {o["ods_code"]: o.get("parent_pcn") for o in practices}
    icb_of = {o["ods_code"]: o.get("parent_icb") for o in practices}
    prac_codes = {o["ods_code"] for o in practices}
    print(f"  {len(practices)} practices in DB\n")

    # ---- 2. fetch mean monthly items per practice, per measure ----
    raw = {}  # metric_key -> {practice: mean items}
    for key_m, codes in MEASURES.items():
        agg = {}
        for c in codes:
            for p, items in mean_items_by_practice(c, months).items():
                agg[p] = agg.get(p, 0) + items
        raw[key_m] = agg
        print(f"  {key_m:<16} {len(agg)} practices with items")

    # statin total + high-intensity share
    statin = {}
    for c in STATIN_CHEMS:
        for p, items in mean_items_by_practice(c, months).items():
            statin[p] = statin.get(p, 0) + items
    print(f"  rx_statin        {len(statin)} practices with items")
    hi = {}
    for c in hi_statin_presentation_codes():
        for p, items in mean_items_by_practice(c, months).items():
            hi[p] = hi.get(p, 0) + items
    print(f"  rx_statin_hi     {len(hi)} practices (high-intensity presentations)\n")

    # ratio (%) measures: numerator group vs a denominator group
    warfarin = {}
    for c in WARFARIN_CODES:
        for p, items in mean_items_by_practice(c, months).items():
            warfarin[p] = warfarin.get(p, 0) + items
    metformin_all = mean_items_by_practice(METFORMIN_CHEM, months)
    mr_codes = [it["id"] for it in (api_get("/bnf_code/", {"q": METFORMIN_CHEM}) or [])
                if len(it.get("id", "")) == 15 and it["id"].startswith(METFORMIN_CHEM)
                and any(h in (" " + (it.get("name", "").lower()) + " ") for h in MR_NAME_HINTS)]
    metformin_mr = {}
    for c in mr_codes:
        for p, items in mean_items_by_practice(c, months).items():
            metformin_mr[p] = metformin_mr.get(p, 0) + items
    print(f"  rx_doac_anticoag warfarin from {len(warfarin)} practices")
    print(f"  rx_metformin_mr  {len(mr_codes)} MR presentations, {len(metformin_mr)} practices\n")

    # ---- 3. build rows (practice + rollups + percentile + derived decile) ----
    rows = []

    def add_ratio(metric_key, num_by, den_by):
        """A % measure = numerator items / denominator items * 100 (higher = better)."""
        def rate_for(codes_num, codes_den):  # helper on aggregated dicts
            pass
        # practice
        prate, praw = {}, {}
        for p in practices:
            c = p["ods_code"]; d = den_by.get(c, 0)
            praw[c] = num_by.get(c, 0)
            prate[c] = round(num_by.get(c, 0) / d * 1000) / 10 if d > 0 else None
        pc = percentiles(prate)
        for p in practices:
            c = p["ods_code"]
            if prate[c] is None:
                continue
            rows.append({"ods_code": c, "org_level": "practice", "metric_key": metric_key, "period": period,
                         "raw_items": praw[c], "items_per_1000": prate[c],
                         "percentile": pc[c], "decile": min(10, max(1, pc[c] // 10 + 1))})
        # rollups + england (list-weighted by summing num/den items)
        for level, parent in (("pcn", pcn_of), ("icb", icb_of)):
            num, den = {}, {}
            for p in practices:
                c = p["ods_code"]; par = parent.get(c)
                if not par:
                    continue
                num[par] = num.get(par, 0) + num_by.get(c, 0)
                den[par] = den.get(par, 0) + den_by.get(c, 0)
            rate = {par: (round(num[par] / den[par] * 1000) / 10 if den[par] > 0 else None) for par in num}
            lp = percentiles(rate)
            for par, rv in rate.items():
                if rv is None:
                    continue
                rows.append({"ods_code": par, "org_level": level, "metric_key": metric_key, "period": period,
                             "raw_items": num[par], "items_per_1000": rv,
                             "percentile": lp[par], "decile": min(10, max(1, lp[par] // 10 + 1))})
        tn = sum(num_by.get(p["ods_code"], 0) for p in practices)
        td = sum(den_by.get(p["ods_code"], 0) for p in practices)
        if td > 0:
            rows.append({"ods_code": "ENG", "org_level": "national", "metric_key": metric_key, "period": period,
                         "raw_items": tn, "items_per_1000": round(tn / td * 1000) / 10, "percentile": 50, "decile": 5})

    def add_level(metric_key, is_share, prate, praw):
        pct = percentiles(prate)
        for p in practices:
            code = p["ods_code"]
            if prate.get(code) is None:
                continue
            pc = pct[code]
            rows.append({"ods_code": code, "org_level": "practice", "metric_key": metric_key, "period": period,
                         "raw_items": praw.get(code), "items_per_1000": prate[code],
                         "percentile": pc, "decile": min(10, max(1, pc // 10 + 1))})
        # rollups
        for level, parent in (("pcn", pcn_of), ("icb", icb_of)):
            num, den, hinum, stnum = {}, {}, {}, {}
            for p in practices:
                code = p["ods_code"]; par = parent.get(code)
                if not par:
                    continue
                if is_share:
                    hinum[par] = hinum.get(par, 0) + hi.get(code, 0)
                    stnum[par] = stnum.get(par, 0) + statin.get(code, 0)
                elif praw.get(code) is not None:
                    num[par] = num.get(par, 0) + praw[code]
                    den[par] = den.get(par, 0) + list_by.get(code, 0)
            rate = {}
            for par in (stnum if is_share else num):
                if is_share:
                    rate[par] = round(hinum[par] / stnum[par] * 1000) / 10 if stnum[par] > 0 else None
                else:
                    rate[par] = round(num[par] / den[par] * 1000, 2) if den[par] > 0 else None
            lp = percentiles(rate)
            for par, rv in rate.items():
                if rv is None or par not in prac_codes and par not in list_by:
                    pass
                if rv is None:
                    continue
                pc = lp[par]
                rows.append({"ods_code": par, "org_level": level, "metric_key": metric_key, "period": period,
                             "raw_items": (hinum if is_share else num).get(par), "items_per_1000": rv,
                             "percentile": pc, "decile": min(10, max(1, pc // 10 + 1))})
        # England
        if is_share:
            hs = sum(hi.get(p["ods_code"], 0) for p in practices)
            ss = sum(statin.get(p["ods_code"], 0) for p in practices)
            eng = round(hs / ss * 1000) / 10 if ss > 0 else None
            eraw = hs
        else:
            ns = sum(praw.get(p["ods_code"], 0) or 0 for p in practices)
            ls = sum(list_by.get(p["ods_code"], 0) for p in practices if praw.get(p["ods_code"]) is not None)
            eng = round(ns / ls * 1000, 2) if ls > 0 else None
            eraw = ns
        if eng is not None:
            rows.append({"ods_code": "ENG", "org_level": "national", "metric_key": metric_key, "period": period,
                         "raw_items": eraw, "items_per_1000": eng, "percentile": 50, "decile": 5})

    for key_m in MEASURES:
        prate = {p["ods_code"]: (round(raw[key_m].get(p["ods_code"], 0) / list_by[p["ods_code"]] * 1000, 2)
                                 if list_by.get(p["ods_code"]) and raw[key_m].get(p["ods_code"]) is not None else None)
                 for p in practices}
        add_level(key_m, False, prate, raw[key_m])
    # statin volume
    prate = {p["ods_code"]: (round(statin.get(p["ods_code"], 0) / list_by[p["ods_code"]] * 1000, 2)
                             if list_by.get(p["ods_code"]) and p["ods_code"] in statin else None) for p in practices}
    add_level("rx_statin", False, prate, statin)
    # HI statin share
    srate = {p["ods_code"]: (round(hi.get(p["ods_code"], 0) / statin[p["ods_code"]] * 1000) / 10
                             if statin.get(p["ods_code"]) else None) for p in practices}
    add_level("rx_statin_hi", True, srate, hi)

    # ratio (%) measures — higher is better
    doac_total = raw["rx_doac"]  # already summed DOAC items
    add_ratio("rx_doac_anticoag", doac_total, {p["ods_code"]: doac_total.get(p["ods_code"], 0) + warfarin.get(p["ods_code"], 0) for p in practices})
    add_ratio("rx_metformin_mr", metformin_mr, metformin_all)

    print(f"Built {len(rows)} rows.")
    if DRY:
        eng = {r["metric_key"]: r["items_per_1000"] for r in rows if r["org_level"] == "national"}
        print("England rates:", eng)
        print("--dry: nothing written.")
        return
    print("Writing to rx_value...")
    sb_upsert(url, key, "rx_value", rows)
    print(f"Done — {len(rows)} rows for {period}.")


if __name__ == "__main__":
    main()
